routing/turns_generator.cpp

```diff
@@ -437,68 +443,162 @@ namespace routing
-  vector<SingleLaneInfo> lanes;
-  if (seg1.IsValid())
+  for (size_t k = startK; k < endK; ++k)
   {
-    FeatureType ft1;
-    Index::FeaturesLoaderGuard loader1(index, routingMapping.GetMwmId());
-    loader1.GetFeatureByIndex(seg1.m_fid, ft1);
-
-    using feature::Metadata;
-    ft1.ParseMetadata();
-    Metadata const & md = ft1.GetMetadata();
+    auto const & segment = buffer[k];
+    if (!segment.IsValid())
+    {
+      m_path.clear();
+      return;
+    }
+    // Load data from drive.
+    FeatureType ft;
+    Index::FeaturesLoaderGuard loader(index, mapping.GetMwmId());
+    loader.GetFeatureByIndex(segment.m_fid, ft);
+    ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
 
-    if (ftypes::IsOneWayChecker::Instance()(ft1))
+    // Get points in proper direction.
+    auto startIdx = segment.m_pointStart;
+    auto endIdx = segment.m_pointEnd;
+    if (isStartNode && k == startK && startGraphNode.segment.IsValid())
+      startIdx = (segment.m_pointEnd > segment.m_pointStart) ? startGraphNode.segment.m_pointStart
+                                                             : startGraphNode.segment.m_pointEnd;
+    if (isEndNode && k == endK - 1 && endGraphNode.segment.IsValid())
+      endIdx = (segment.m_pointEnd > segment.m_pointStart) ? endGraphNode.segment.m_pointEnd
+                                                           : endGraphNode.segment.m_pointStart;
+    if (startIdx < endIdx)
     {
-      ParseLanes(md.Get(Metadata::FMD_TURN_LANES), lanes);
-      return lanes;
+      for (auto idx = startIdx; idx <= endIdx; ++idx)
+        m_path.push_back(ft.GetPoint(idx));
     }
-    // two way roads
-    if (seg1.m_pointStart < seg1.m_pointEnd)
+    else
+    {
+      // I use big signed type because endIdx can be 0.
+      for (int64_t idx = startIdx; idx >= endIdx; --idx)
+        m_path.push_back(ft.GetPoint(idx));
+    }
+
+    // Load lanes if it is a last segment before junction.
+    if (buffer.back() == segment)
     {
-      // forward direction
-      ParseLanes(md.Get(Metadata::FMD_TURN_LANES_FORWARD), lanes);
-      return lanes;
+      using feature::Metadata;
+      ft.ParseMetadata();
+      Metadata const & md = ft.GetMetadata();
+
+      auto directionType = Metadata::FMD_TURN_LANES;
+
+      if (!ftypes::IsOneWayChecker::Instance()(ft))
+      {
+        directionType = (startIdx < endIdx) ? Metadata::FMD_TURN_LANES_FORWARD
+                                            : Metadata::FMD_TURN_LANES_BACKWARD;
+      }
+      ParseLanes(md.Get(directionType), m_lanes);
     }
-    // backward direction
-    ParseLanes(md.Get(Metadata::FMD_TURN_LANES_BACKWARD), lanes);
-    return lanes;
+    // Calculate node flags.
+    m_onRoundabout |= ftypes::IsRoundAboutChecker::Instance()(ft);
+    m_isLink |= ftypes::IsLinkChecker::Instance()(ft);
+    m_highwayClass = ftypes::GetHighwayClass(ft);
+    string name;
+    ft.GetName(FeatureType::DEFAULT_LANG, name);
+    if (!name.empty())
```

Comment:
Because I assume cases when there will be segments with empty names. They must not corrupt valid road name.